app/src/processing/app/EditorConsole.java

```diff
@@ -179,10 +185,48 @@ public boolean isEmpty() {
     return document.getLength() == 0;
   }
 
-  public void insertString(String line, SimpleAttributeSet attributes) throws BadLocationException {
-    line = line.replace("\r\n", "\n").replace("\r", "\n");
-    int offset = document.getLength();
-    document.insertString(offset, line, attributes);
+  public void insertString(String str, SimpleAttributeSet attributes) throws BadLocationException {
+    // Separate the string into content, newlines and lone carriage
+    // returns.
+    //
+    // Doing so allows lone CRs to return the insertPosition to the
+    // start of the line to allow overwriting the most recent line (e.g.
+    // for a progress bar). Any CR or NL that are immediately followed
+    // by another NL are bunched together for efficiency, since these
+    // can just be inserted into the document directly and still be
+    // correct.
+    //
+    // This regex is written so it will necessarily match any string
+    // completely if applied repeatedly. This is important because any
+    // part not matched would be silently dropped.
+    Matcher m = Pattern.compile("([^\r\n]*)([\r\n]*\n)?(\r+)?").matcher(str);
```

Comment:
Knowing this will be called for each and every line printed to the console, I wonder if it wouldn't make sense to move the time consuming compiled pattern to a static variable at the top of the file:
`private static Pattern newLinePattern = Pattern.compile("([^\r\n]*)([\r\n]*\n)?(\r+)?");`
Then here in the insertString method, just call 
`Matcher m = newLinePattern.matcher(str);`
Maybe premature optimization is the root of all evil, but repeatedly compiling the same Pattern can be quite expensive...

Reply:
Good idea, I'll see about that soon.

Reply:
Done, just pushed this.